include/query/algorithm/exhaustive_taat_query.hpp

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include "util/intrinsics.hpp"
+#include "topk_queue.hpp"
+
+namespace pisa {
+
+// TODO: These are functions common to query processing in general.
+//       They should be moved out of this file.
+namespace query {
+
+template <typename Index, typename WandType>
+[[nodiscard]] auto cursors_with_scores(Index const& index, WandType const &wdata, term_id_vec terms)
+{
+    // TODO(michal): parametrize scorer_type; didn't do that because this might mean some more
+    //               complex refactoring I want to avoid for now.
+    using scorer_type         = bm25;
+    using cursor_type         = typename Index::document_enumerator;
+    using score_function_type = std::function<float(uint64_t, uint64_t)>;
+
+    auto query_term_freqs = query_freqs(terms);
+    std::vector<cursor_type> cursors;
+    std::vector<score_function_type> score_functions;
+    cursors.reserve(query_term_freqs.size());
+    score_functions.reserve(query_term_freqs.size());
+
+    for (auto term : query_term_freqs) {
+        auto     list     = index[term.first];
+        uint64_t num_docs = index.num_docs();
+        auto     q_weight = scorer_type::query_term_weight(term.second, list.size(), num_docs);
+        cursors.push_back(std::move(list));
+        score_functions.push_back([q_weight, &wdata](auto docid, auto freq) {
+            float norm_len = wdata.norm_len(docid);
+            return q_weight * scorer_type::doc_term_weight(freq, norm_len);
+        });
+    }
+    return std::make_pair(cursors, score_functions);
+}
+
+} // namespace query
+
+template <typename Index, typename WandType, size_t accumulator_block_size = 0>
+struct exhaustive_taat_query {
+    exhaustive_taat_query(Index const &index, WandType const &wdata, uint64_t k)
+        : m_index(index), m_wdata(wdata), m_topk(k), m_accumulators(index.num_docs())
+    {
+        static_assert(accumulator_block_size >= 0, "must be non-negative");
```

Comment:
I guess. I think I'm just not a big fan of unsigned integers for arithmetic. But I used it coz all the other sizes seem to be unsigned, and I left the assert.

Reply:
Actually, it **should** be a signed integer. Just think about it. Someone could very well declare `exhaustive_taat_query<.., .., -1>` thinking it's "no blocks" or by accident. It will compile and define the block size to a huge number. That's just weird and unexpected. That's precisely why you shouldn't use unsigned integers unless you do bit manipulation. Sadly, someone thought a while ago that a container size is `size_t`, which is stupid.